File: enode/src/main/java/org/enodeframework/eventing/impl/DefaultProcessingEventProcessor.java
package org.enodeframework.eventing.impl;

import com.google.common.base.Strings;
import com.google.common.collect.Lists;
import org.enodeframework.common.io.IOHelper;
import org.enodeframework.common.io.Task;
import org.enodeframework.common.scheduling.IScheduleService;
import org.enodeframework.common.serializing.ISerializeService;
import org.enodeframework.eventing.DomainEventStreamMessage;
import org.enodeframework.eventing.EnqueueMessageResult;
import org.enodeframework.eventing.IProcessingEventProcessor;
import org.enodeframework.eventing.IPublishedVersionStore;
import org.enodeframework.eventing.ProcessingEvent;
import org.enodeframework.eventing.ProcessingEventMailBox;
import org.enodeframework.messaging.IMessageDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executor;
import java.util.stream.Collectors;

/**
 * @author anruence@gmail.com
 */
public class DefaultProcessingEventProcessor implements IProcessingEventProcessor {

    private static final Logger logger = LoggerFactory.getLogger(DefaultProcessingEventProcessor.class);
    private final String scanInactiveMailBoxTaskName;
    private final String processTryToRefreshAggregateTaskName;
    private final String name = "DefaultEventProcessor";
    private final ConcurrentHashMap<String, ProcessingEventMailBox> toRefreshAggregateRootMailBoxDict;
    private final ConcurrentHashMap<String, ProcessingEventMailBox> mailboxDict;
    private final ConcurrentHashMap<String, Boolean> refreshingAggregateRootDict;
    private final IScheduleService scheduleService;
    private final ISerializeService serializeService;
    private final IMessageDispatcher messageDispatcher;
    private final IPublishedVersionStore publishedVersionStore;
    private final Executor executor;
    private int timeoutSeconds = 3600 * 24 * 3;
    private int scanExpiredAggregateIntervalMilliseconds = 5000;
    private int processTryToRefreshAggregateIntervalMilliseconds = 1000;

    public DefaultProcessingEventProcessor(IScheduleService scheduleService, ISerializeService serializeService, IMessageDispatcher messageDispatcher, IPublishedVersionStore publishedVersionStore, Executor executor) {
        this.scheduleService = scheduleService;
        this.serializeService = serializeService;
        this.messageDispatcher = messageDispatcher;
        this.publishedVersionStore = publishedVersionStore;
        this.executor = executor;
        this.mailboxDict = new ConcurrentHashMap<>();
        this.toRefreshAggregateRootMailBoxDict = new ConcurrentHashMap<>();
        this.refreshingAggregateRootDict = new ConcurrentHashMap<>();
        this.scanInactiveMailBoxTaskName = "CleanInactiveProcessingEventMailBoxes_" + System.currentTimeMillis() + new Random().nextInt(10000);
        this.processTryToRefreshAggregateTaskName = "ProcessTryToRefreshAggregate_" + System.currentTimeMillis() + new Random().nextInt(10000);
    }

    @Override
    public void process(ProcessingEvent processingMessage) {
        String aggregateRootId = processingMessage.getMessage().getAggregateRootId();
        if (Strings.isNullOrEmpty(aggregateRootId)) {
            throw new IllegalArgumentException("aggregateRootId of domain event stream cannot be null or empty, domainEventStreamId:" + processingMessage.getMessage().getId());
        }
        ProcessingEventMailBox mailbox = mailboxDict.computeIfAbsent(aggregateRootId, key -> buildProcessingEventMailBox(processingMessage));
        long mailboxTryUsingCount = 0L;
        while (!mailbox.tryUsing()) {
            Task.sleep(1);
            mailboxTryUsingCount++;
            if (mailboxTryUsingCount % 10000 == 0) {
                logger.warn("Event mailbox try using count: {}, aggregateRootId: {}, aggregateRootTypeName: {}", mailboxTryUsingCount, mailbox.getAggregateRootId(), mailbox.getAggregateRootTypeName());
            }
        }
        if (mailbox.isRemoved()) {
            mailbox = buildProcessingEventMailBox(processingMessage);
            mailboxDict.putIfAbsent(aggregateRootId, mailbox);
        }
        EnqueueMessageResult enqueueResult = mailbox.enqueueMessage(processingMessage);
        if (enqueueResult == EnqueueMessageResult.Ignored) {
            processingMessage.getProcessContext().notifyEventProcessed();
        } else if (enqueueResult == EnqueueMessageResult.AddToWaitingList) {
            addToRefreshAggregateMailBoxToDict(mailbox);
        }
        mailbox.exitUsing();
    }

    private void addToRefreshAggregateMailBoxToDict(ProcessingEventMailBox mailbox) {
        if (toRefreshAggregateRootMailBoxDict.putIfAbsent(mailbox.getAggregateRootId(), mailbox) == null) {
            logger.info("Added toRefreshPublishedVersion aggregate mailbox, aggregateRootTypeName: {}, aggregateRootId: {}", mailbox.getAggregateRootTypeName(), mailbox.getAggregateRootId());
            tryToRefreshAggregateMailBoxNextExpectingEventVersion(mailbox);
        }
    }

    private ProcessingEventMailBox buildProcessingEventMailBox(ProcessingEvent processingMessage) {
        return new ProcessingEventMailBox(processingMessage.getMessage().getAggregateRootTypeName(), processingMessage.getMessage().getAggregateRootId(), y -> dispatchProcessingMessageAsync(y, 0), executor);
    }

    private void tryToRefreshAggregateMailBoxNextExpectingEventVersion(ProcessingEventMailBox processingEventMailBox) {
        if (refreshingAggregateRootDict.putIfAbsent(processingEventMailBox.getAggregateRootId(), true) == null) {
            getAggregateRootLatestPublishedEventVersion(processingEventMailBox, 0);
        }
    }

    private void getAggregateRootLatestPublishedEventVersion(ProcessingEventMailBox processingEventMailBox, int retryTimes) {
        IOHelper.tryAsyncActionRecursively("GetAggregateRootLatestPublishedEventVersion",
                () -> publishedVersionStore.getPublishedVersionAsync(name, processingEventMailBox.getAggregateRootTypeName(), processingEventMailBox.getAggregateRootId()),
                result -> {
                    processingEventMailBox.setNextExpectingEventVersion(result + 1);
                    refreshingAggregateRootDict.remove(processingEventMailBox.getAggregateRootId());
                },
                () -> String.format("publishedVersionStore.GetPublishedVersionAsync has unknown exception, aggregateRootTypeName: %s, aggregateRootId: %s", processingEventMailBox.getAggregateRootTypeName(), processingEventMailBox.getAggregateRootId()),
                null,
                retryTimes,
                true);
    }

    @Override
    public void start() {
        scheduleService.startTask(scanInactiveMailBoxTaskName, this::cleanInactiveMailbox, scanExpiredAggregateIntervalMilliseconds, scanExpiredAggregateIntervalMilliseconds);
        scheduleService.startTask(processTryToRefreshAggregateTaskName, this::processToRefreshAggregateRootMailBoxs, processTryToRefreshAggregateIntervalMilliseconds, processTryToRefreshAggregateIntervalMilliseconds);
    }

    @Override
    public void stop() {
        scheduleService.stopTask(scanInactiveMailBoxTaskName);
        scheduleService.stopTask(processTryToRefreshAggregateTaskName);
    }

    private void dispatchProcessingMessageAsync(ProcessingEvent processingEvent, int retryTimes) {
        IOHelper.tryAsyncActionRecursivelyWithoutResult("DispatchProcessingMessageAsync",
                () -> messageDispatcher.dispatchMessagesAsync(processingEvent.getMessage().getEvents()),
                result -> {
                    if (logger.isDebugEnabled()) {
                        logger.debug("dispatch messages success, msg: {}", serializeService.serialize(processingEvent.getMessage()));
                    }
                    updatePublishedVersionAsync(processingEvent, 0);
                },
                () -> String.format("sequence message [messageId:%s, messageType:%s, aggregateRootId:%s, aggregateRootVersion:%s]", processingEvent.getMessage().getId(), processingEvent.getMessage().getClass().getName(), processingEvent.getMessage().getAggregateRootId(), processingEvent.getMessage().getVersion()),
                null,
                retryTimes, true);
    }

    /**
     * The name of the processor
     */
    @Override
    public String getName() {
        return name;
    }

    private void updatePublishedVersionAsync(ProcessingEvent processingEvent, int retryTimes) {
        DomainEventStreamMessage message = processingEvent.getMessage();
        IOHelper.tryAsyncActionRecursivelyWithoutResult("UpdatePublishedVersionAsync",
                () -> {
                    return publishedVersionStore.updatePublishedVersionAsync(name, message.getAggregateRootTypeName(), message.getAggregateRootId(), message.getVersion());
                },
                result -> {
                    if (logger.isDebugEnabled()) {
                        logger.debug("update published version success, message ack: {}", serializeService.serialize(message));
                    }
                    processingEvent.complete();
                },
                () -> String.format("DomainEventStreamMessage [messageId:%s, messageType:%s, aggregateRootId:%s, aggregateRootVersion:%s]", message.getId(), message.getClass().getName(), message.getAggregateRootId(), message.getVersion()),
                null, retryTimes, true);
    }

    private void processToRefreshAggregateRootMailBoxs() {
        List<ProcessingEventMailBox> remainingMailboxList = Lists.newArrayList();
        List<ProcessingEventMailBox> recoveredMailboxList = Lists.newArrayList();
        toRefreshAggregateRootMailBoxDict.values().forEach(aggregateRootMailBox -> {
            if (aggregateRootMailBox.getWaitingMessageCount() > 0) {
                remainingMailboxList.add(aggregateRootMailBox);
            } else {
                recoveredMailboxList.add(aggregateRootMailBox);
            }
        });
        for (ProcessingEventMailBox mailBox : remainingMailboxList) {
            tryToRefreshAggregateMailBoxNextExpectingEventVersion(mailBox);
        }
        for (ProcessingEventMailBox mailBox : recoveredMailboxList) {
            ProcessingEventMailBox removed = toRefreshAggregateRootMailBoxDict.remove(mailBox.getAggregateRootId());
            if (removed != null) {
                logger.info("Removed healthy aggregate mailbox, aggregateRootTypeName: {}, aggregateRootId: {}", removed.getAggregateRootTypeName(), removed.getAggregateRootId());
            }
        }
    }

    private void cleanInactiveMailbox() {
        List<Map.Entry<String, ProcessingEventMailBox>> inactiveList = mailboxDict.entrySet().stream()
                .filter(x -> isMailBoxAllowRemove(x.getValue()))
                .collect(Collectors.toList());
        inactiveList.forEach(entry -> {
            if (entry.getValue().tryUsing()) {
                if (isMailBoxAllowRemove(entry.getValue())) {
                    ProcessingEventMailBox removed = mailboxDict.remove(entry.getKey());
                    if (removed != null) {
                        removed.markAsRemoved();
                        logger.info("Removed inactive domain event stream mailbox, aggregateRootTypeName: {}, aggregateRootId: {}", removed.getAggregateRootTypeName(), removed.getAggregateRootId());
                    }
                }
            }
        });
    }

    private boolean isMailBoxAllowRemove(ProcessingEventMailBox mailbox) {
        return mailbox.isInactive(timeoutSeconds)
                && !mailbox.isRunning()
                && mailbox.getTotalUnHandledMessageCount() == 0
                && mailbox.getWaitingMessageCount() == 0;
    }

    public int getTimeoutSeconds() {
        return timeoutSeconds;
    }

    public void setTimeoutSeconds(int timeoutSeconds) {
        this.timeoutSeconds = timeoutSeconds;
    }

    public int getScanExpiredAggregateIntervalMilliseconds() {
        return scanExpiredAggregateIntervalMilliseconds;
    }

    public void setScanExpiredAggregateIntervalMilliseconds(int scanExpiredAggregateIntervalMilliseconds) {
        this.scanExpiredAggregateIntervalMilliseconds = scanExpiredAggregateIntervalMilliseconds;
    }

    public int getProcessTryToRefreshAggregateIntervalMilliseconds() {
        return processTryToRefreshAggregateIntervalMilliseconds;
    }

    public void setProcessTryToRefreshAggregateIntervalMilliseconds(int processTryToRefreshAggregateIntervalMilliseconds) {
        this.processTryToRefreshAggregateIntervalMilliseconds = processTryToRefreshAggregateIntervalMilliseconds;
    }
}

File: enode/src/main/java/org/enodeframework/eventing/impl/DefaultProcessingEventProcessor.kt
package org.enodeframework.eventing.impl

import com.google.common.base.Strings
import com.google.common.collect.Lists
import kotlinx.coroutines.CoroutineDispatcher
import org.enodeframework.common.extensions.SystemClock
import org.enodeframework.common.io.IOHelper.tryAsyncActionRecursively
import org.enodeframework.common.io.IOHelper.tryAsyncActionRecursivelyWithoutResult
import org.enodeframework.common.io.Task.sleep
import org.enodeframework.common.scheduling.ScheduleService
import org.enodeframework.common.serializing.SerializeService
import org.enodeframework.eventing.*
import org.enodeframework.messaging.MessageDispatcher
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * @author anruence@gmail.com
 */
class DefaultProcessingEventProcessor(
    private val scheduleService: ScheduleService,
    private val serializeService: SerializeService,
    private val messageDispatcher: MessageDispatcher,
    private val publishedVersionStore: PublishedVersionStore,
    private val coroutineDispatcher: CoroutineDispatcher
) : ProcessingEventProcessor {
    private val scanInactiveMailBoxTaskName: String =
        "CleanInactiveProcessingEventMailBoxes_" + SystemClock.now() + Random().nextInt(10000)
    private val processTryToRefreshAggregateTaskName: String =
        "ProcessTryToRefreshAggregate_" + SystemClock.now() + Random().nextInt(10000)

    /**
     * The name of the processor
     */
    override val name = "DefaultEventProcessor"
    private val toRefreshAggregateRootMailBoxDict: ConcurrentHashMap<String, ProcessingEventMailBox> =
        ConcurrentHashMap()
    private val mailboxDict: ConcurrentHashMap<String, ProcessingEventMailBox> = ConcurrentHashMap()
    private val refreshingAggregateRootDict: ConcurrentHashMap<String, Boolean> = ConcurrentHashMap()
    private var timeoutSeconds = 3600 * 24 * 3
    private var scanExpiredAggregateIntervalMilliseconds = 5000
    private var processTryToRefreshAggregateIntervalMilliseconds = 1000

    override fun process(processingEvent: ProcessingEvent) {
        val aggregateRootId = processingEvent.message.getAggregateRootId()
        require(!Strings.isNullOrEmpty(aggregateRootId)) { "aggregateRootId of domain event stream cannot be null or empty, domainEventStreamId:" + processingEvent.message.id }
        var mailbox = mailboxDict.computeIfAbsent(aggregateRootId) { buildProcessingEventMailBox(processingEvent) }
        var mailboxTryUsingCount = 0L
        while (!mailbox.tryUsing()) {
            sleep(1)
            mailboxTryUsingCount++
            if (mailboxTryUsingCount % 10000 == 0L) {
                logger.warn(
                    "Event mailbox try using count: {}, aggregateRootId: {}, aggregateRootTypeName: {}",
                    mailboxTryUsingCount,
                    mailbox.aggregateRootId,
                    mailbox.aggregateRootTypeName
                )
            }
        }
        if (mailbox.isRemoved()) {
            mailbox = mailboxDict.computeIfAbsent(aggregateRootId) {
                buildProcessingEventMailBox(processingEvent)
            }
        }
        val enqueueResult = mailbox.enqueueMessage(processingEvent)
        if (enqueueResult == EnqueueMessageResult.Ignored) {
            processingEvent.processContext.notifyEventProcessed()
        } else if (enqueueResult == EnqueueMessageResult.AddToWaitingList) {
            addToRefreshAggregateMailBoxToDict(mailbox)
        }
        mailbox.exitUsing()
    }

    private fun addToRefreshAggregateMailBoxToDict(mailbox: ProcessingEventMailBox) {
        if (toRefreshAggregateRootMailBoxDict.putIfAbsent(mailbox.aggregateRootId, mailbox) == null) {
            logger.info(
                "Added toRefreshPublishedVersion aggregate mailbox, aggregateRootTypeName: {}, aggregateRootId: {}",
                mailbox.aggregateRootTypeName,
                mailbox.aggregateRootId
            )
            tryToRefreshAggregateMailBoxNextExpectingEventVersion(mailbox)
        }
    }

    private fun buildProcessingEventMailBox(processingMessage: ProcessingEvent): ProcessingEventMailBox {
        return ProcessingEventMailBox(
            processingMessage.message.aggregateRootTypeName,
            processingMessage.message.aggregateRootId,
            coroutineDispatcher
        ) { y: ProcessingEvent -> dispatchProcessingMessageAsync(y, 0) }
    }

    private fun tryToRefreshAggregateMailBoxNextExpectingEventVersion(processingEventMailBox: ProcessingEventMailBox) {
        if (refreshingAggregateRootDict.putIfAbsent(processingEventMailBox.aggregateRootId, true) == null) {
            getAggregateRootLatestPublishedEventVersion(processingEventMailBox, 0)
        }
    }

    private fun getAggregateRootLatestPublishedEventVersion(
        processingEventMailBox: ProcessingEventMailBox, retryTimes: Int
    ) {
        tryAsyncActionRecursively("GetAggregateRootLatestPublishedEventVersion", {
            publishedVersionStore.getPublishedVersionAsync(
                name, processingEventMailBox.aggregateRootTypeName, processingEventMailBox.aggregateRootId
            )
        }, { result: Int ->
            processingEventMailBox.setNextExpectingEventVersion(result + 1)
            refreshingAggregateRootDict.remove(processingEventMailBox.aggregateRootId)
        }, {
            String.format(
                "publishedVersionStore.GetPublishedVersionAsync has unknown exception, aggregateRootTypeName: %s, aggregateRootId: %s",
                processingEventMailBox.aggregateRootTypeName,
                processingEventMailBox.aggregateRootId
            )
        }, null, retryTimes, true
        )
    }

    override fun start() {
        scheduleService.startTask(
            scanInactiveMailBoxTaskName,
            { cleanInactiveMailbox() },
            scanExpiredAggregateIntervalMilliseconds,
            scanExpiredAggregateIntervalMilliseconds
        )
        scheduleService.startTask(
            processTryToRefreshAggregateTaskName,
            { processToRefreshAggregateRootMailBoxs() },
            processTryToRefreshAggregateIntervalMilliseconds,
            processTryToRefreshAggregateIntervalMilliseconds
        )
    }

    override fun stop() {
        scheduleService.stopTask(scanInactiveMailBoxTaskName)
        scheduleService.stopTask(processTryToRefreshAggregateTaskName)
    }

    private fun dispatchProcessingMessageAsync(processingEvent: ProcessingEvent, retryTimes: Int) {
        tryAsyncActionRecursivelyWithoutResult(
            "DispatchProcessingMessageAsync",
            { messageDispatcher.dispatchMessagesAsync(processingEvent.message.events) },
            {
                if (logger.isDebugEnabled) {
                    logger.debug(
                        "dispatch messages success, msg: {}", serializeService.serialize(processingEvent.message)
                    )
                }
                updatePublishedVersionAsync(processingEvent, 0)
            },
            {
                String.format(
                    "sequence message [messageId:%s, messageType:%s, aggregateRootId:%s, aggregateRootVersion:%s]",
                    processingEvent.message.id,
                    processingEvent.message.javaClass.name,
                    processingEvent.message.getAggregateRootId(),
                    processingEvent.message.getVersion()
                )
            },
            null,
            retryTimes,
            true
        )
    }

    private fun updatePublishedVersionAsync(processingEvent: ProcessingEvent, retryTimes: Int) {
        val message = processingEvent.message
        tryAsyncActionRecursivelyWithoutResult("UpdatePublishedVersionAsync", {
            publishedVersionStore.updatePublishedVersionAsync(
                name, message.getAggregateRootTypeName(), message.getAggregateRootId(), message.getVersion()
            )
        }, {
            if (logger.isDebugEnabled) {
                logger.debug(
                    "update published version success, message ack: {}", serializeService.serialize(message)
                )
            }
            processingEvent.complete()
        }, {
            String.format(
                "DomainEventStreamMessage [messageId:%s, messageType:%s, aggregateRootId:%s, aggregateRootVersion:%s]",
                message.id,
                message.javaClass.name,
                message.getAggregateRootId(),
                message.getVersion()
            )
        }, null, retryTimes, true
        )
    }

    private fun processToRefreshAggregateRootMailBoxs() {
        val remainingMailboxList: MutableList<ProcessingEventMailBox> = Lists.newArrayList()
        val recoveredMailboxList: MutableList<ProcessingEventMailBox> = Lists.newArrayList()
        toRefreshAggregateRootMailBoxDict.values.forEach { aggregateRootMailBox: ProcessingEventMailBox ->
            if (aggregateRootMailBox.getWaitingMessageCount() > 0) {
                remainingMailboxList.add(aggregateRootMailBox)
            } else {
                recoveredMailboxList.add(aggregateRootMailBox)
            }
        }
        for (mailBox in remainingMailboxList) {
            tryToRefreshAggregateMailBoxNextExpectingEventVersion(mailBox)
        }
        for (mailBox in recoveredMailboxList) {
            val removed = toRefreshAggregateRootMailBoxDict.remove(mailBox.aggregateRootId)
            if (removed != null) {
                logger.info(
                    "Removed healthy aggregate mailbox, aggregateRootTypeName: {}, aggregateRootId: {}",
                    removed.aggregateRootTypeName,
                    removed.aggregateRootId
                )
            }
        }
    }

    private fun cleanInactiveMailbox() {
        val inactiveList = mailboxDict.entries.filter { entry -> isMailBoxAllowRemove(entry.value) }
        inactiveList.forEach { (key, value): Map.Entry<String, ProcessingEventMailBox> ->
            if (value.tryUsing()) {
                if (isMailBoxAllowRemove(value)) {
                    val removed = mailboxDict.remove(key)
                    if (removed != null) {
                        removed.markAsRemoved()
                        logger.info(
                            "Removed inactive domain event stream mailbox, aggregateRootTypeName: {}, aggregateRootId: {}",
                            removed.aggregateRootTypeName,
                            removed.aggregateRootId
                        )
                    }
                }
            }
        }
    }

    private fun isMailBoxAllowRemove(mailbox: ProcessingEventMailBox): Boolean {
        return (mailbox.isInactive(timeoutSeconds) && !mailbox.isRunning() && mailbox.getTotalUnHandledMessageCount() == 0 && mailbox.getWaitingMessageCount() == 0)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DefaultProcessingEventProcessor::class.java)
    }

}